app/util/StepFunctions.scala

```diff
@@ -0,0 +1,120 @@
+      val json = Json.parse(result.getInput)
+
+      Some(json.as[Upload])
+    } catch {
+      case _: ExecutionDoesNotExistException =>
+        None
+    }
+  }
+
+  def getStatus(atomId: String): Iterable[UploadStatus] = {
+    val runningJobs = getExecutions(atomId, ExecutionStatus.RUNNING)
+    val failedJobs = getExecutions(atomId, ExecutionStatus.FAILED).filter(lessThan10MinutesOld)
+
+    val running = runningJobs.map(getRunningStatus)
+    val failed = failedJobs.map(getFailedStatus)
+
+    running ++ failed
+  }
+
+  def start(upload: Upload): Unit = {
+    val stepFunctionsRequest = new StartExecutionRequest()
+      .withName(upload.id)
+      .withStateMachineArn(awsConfig.pipelineArn)
+      .withInput(Json.stringify(Json.toJson(upload)))
+
+    awsConfig.stepFunctionsClient.startExecution(stepFunctionsRequest)
+  }
+
+  private def getExecutions(atomId: String, filter: ExecutionStatus): Iterable[ExecutionListItem] = {
+    val request = new ListExecutionsRequest()
+      .withStateMachineArn(awsConfig.pipelineArn)
+      .withStatusFilter(filter)
+
+    val results = awsConfig.stepFunctionsClient.listExecutions(request).getExecutions.asScala
+
+    results.filter(_.getName.startsWith(atomId))
+  }
+
+  private def getEvents(execution: ExecutionListItem): Iterable[HistoryEvent] = {
+    val request = new GetExecutionHistoryRequest()
+      .withExecutionArn(execution.getExecutionArn)
+      .withReverseOrder(true)
+      .withMaxResults(20)
+
+    awsConfig.stepFunctionsClient.getExecutionHistory(request).getEvents.asScala
+  }
+
+  private def getRunningStatus(execution: ExecutionListItem): UploadStatus = {
+    val id = execution.getName
+    val events = getEvents(execution)
+
+    events.find(_.getType == "TaskStateEntered") match {
+      case Some(event) =>
+        val state = event.getStateEnteredEventDetails.getName
+        val upload = Json.parse(event.getStateEnteredEventDetails.getInput).as[Upload]
+
+        if(upload.metadata.selfHost) {
+          UploadStatus.indeterminate(id, state)
+        } else {
+          val current = upload.progress.chunksInYouTube
+          val total = upload.parts.length
+
+          if(current < total) {
+            UploadStatus(id, "Uploading to YouTube", current, total, failed = false)
+          } else {
+            UploadStatus.indeterminate(id, state)
+          }
+        }
+
+      case None =>
+        UploadStatus.indeterminate(id, "Uploading")
+    }
+  }
+
+  private def getFailedStatus(execution: ExecutionListItem): UploadStatus = {
+    val id = execution.getName
+    val events = getEvents(execution)
+
+    val cause = events.find(_.getType == "ExecutionFailed") match {
+      case Some(event) =>
+        val cause = event.getExecutionFailedEventDetails.getCause
+
+        try {
+          (Json.parse(cause) \ "errorMessage").as[String]
+        } catch {
+          case _: JsonParseException | _: JsResultException =>
```

Comment:
Not all failures will have a nice error message (lambdas timing out for example).